fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -152,51 +141,67 @@ ACTOR Future<Void> normalizeKeySelectorActor(SpecialKeySpace* sks, ReadYourWrite
-                                                                           GetRangeLimits limits, bool reverse) {
+ACTOR Future<Standalone<RangeResultRef>> SpecialKeySpace::checkRYWValid(SpecialKeySpace* sks,
+                                                                        ReadYourWritesTransaction* ryw,
+                                                                        KeySelector begin, KeySelector end,
+                                                                        GetRangeLimits limits, bool reverse) {
+	ASSERT(ryw);
 	choose {
-		when(std::pair<Standalone<RangeResultRef>, Optional<SpecialKeySpace::MODULE>> result =
+		when(Standalone<RangeResultRef> result =
 		         wait(SpecialKeySpace::getRangeAggregationActor(sks, ryw, begin, end, limits, reverse))) {
-			if (ryw && !ryw->specialKeySpaceRelaxed()) {
-				auto module = result.second;
-				if (!module.present() || module.get() == SpecialKeySpace::MODULE::UNKNOWN) {
-					throw special_keys_no_module_found();
-				}
-			}
-			return result.first;
+			return result;
 		}
-		when(wait(ryw ? ryw->resetFuture() : Never())) { throw internal_error(); }
+		when(wait(ryw->resetFuture())) { throw internal_error(); }
 	}
 }
 
-ACTOR Future<std::pair<Standalone<RangeResultRef>, Optional<SpecialKeySpace::MODULE>>>
-SpecialKeySpace::getRangeAggregationActor(SpecialKeySpace* sks, ReadYourWritesTransaction* ryw, KeySelector begin,
-                                          KeySelector end, GetRangeLimits limits, bool reverse) {
+ACTOR Future<Standalone<RangeResultRef>> SpecialKeySpace::getRangeAggregationActor(SpecialKeySpace* sks,
+                                                                                   ReadYourWritesTransaction* ryw,
+                                                                                   KeySelector begin, KeySelector end,
+                                                                                   GetRangeLimits limits,
+                                                                                   bool reverse) {
 	// This function handles ranges which cover more than one keyrange and aggregates all results
 	// KeySelector, GetRangeLimits and reverse are all handled here
 	state Standalone<RangeResultRef> result;
 	state Standalone<RangeResultRef> pairs;
 	state RangeMap<Key, SpecialKeyRangeBaseImpl*, KeyRangeRef>::Iterator iter;
 	state int actualBeginOffset;
 	state int actualEndOffset;
-	state Optional<SpecialKeySpace::MODULE> lastModuleRead;
+	state KeyRangeRef moduleBoundary;
 	// used to cache result from potential first read
 	state Optional<Standalone<RangeResultRef>> cache;
 
-	wait(normalizeKeySelectorActor(sks, ryw, &begin, &lastModuleRead, &actualBeginOffset, &result, &cache));
-	wait(normalizeKeySelectorActor(sks, ryw, &end, &lastModuleRead, &actualEndOffset, &result, &cache));
+	if (!ryw->specialKeySpaceRelaxed()) {
```

Comment:
Nit: I usually have an easier time parsing this sort of thing if you make the condition just `ryw->specialKeySpaceRelaxed()` and switch the code in each branch

Reply:
Yeah, right. Will switch the condition.